test/algos/adaptors/test_when_all.cpp

```diff
@@ -0,0 +1,328 @@
+  );
+  auto op = ex::connect(std::move(snd), expect_error_receiver{});
+  ex::start(op);
+}
+
+TEST_CASE("when_all terminates with done if one child is cancelled", "[adaptors][when_all]") {
+  done_scheduler sched;
+  ex::sender auto snd = ex::when_all( //
+      ex::just(2),                    //
+      ex::transfer_just(sched, 5),    //
+      ex::just(7)                     //
+  );
+  auto op = ex::connect(std::move(snd), expect_done_receiver{});
+  ex::start(op);
+}
+
+TEST_CASE("TODO: when_all cancels remaining children if error is detected", "[adaptors][when_all]") {
+  impulse_scheduler sched;
+  error_scheduler err_sched;
+  bool called1{false};
+  bool called3{false};
+  bool cancelled{false};
+  ex::sender auto snd = ex::when_all(                                //
+      ex::on(sched, ex::just()) | ex::then([&] { called1 = true; }), //
+      ex::on(sched, ex::transfer_just(err_sched, 5)),                //
+      ex::on(sched, ex::just())                                      //
+          | ex::then([&] { called3 = true; })                        //
+          | ex::let_done([&] {
+              cancelled = true;
+              return ex::just();
+            }) //
+  );
+  auto op = ex::connect(std::move(snd), expect_error_receiver{});
+  ex::start(op);
+  // The first child will complete; the third one will be cancelled
+  CHECK_FALSE(called1);
+  CHECK_FALSE(called3);
+  sched.start_next(); // start the first child
+  CHECK(called1);
+  sched.start_next(); // start the second child; this will generate an error
+  CHECK_FALSE(called3);
+  sched.start_next(); // start the third child
+  // TODO: the third child sender should have been canceled
+  // CHECK_FALSE(called3);
+  // CHECK(cancelled);
+}
+
+TEST_CASE("TODO: when_all cancels remaining children if cancel is detected", "[adaptors][when_all]") {
+  done_scheduler done_sched;
+  impulse_scheduler sched;
+  bool called1{false};
+  bool called3{false};
+  bool cancelled{false};
+  ex::sender auto snd = ex::when_all(                                //
+      ex::on(sched, ex::just()) | ex::then([&] { called1 = true; }), //
+      ex::on(sched, ex::transfer_just(done_sched, 5)),               //
+      ex::on(sched, ex::just())                                      //
+          | ex::then([&] { called3 = true; })                        //
+          | ex::let_done([&] {
+              cancelled = true;
+              return ex::just();
+            }) //
+  );
+  auto op = ex::connect(std::move(snd), expect_done_receiver{});
+  ex::start(op);
+  // The first child will complete; the third one will be cancelled
+  CHECK_FALSE(called1);
+  CHECK_FALSE(called3);
+  sched.start_next(); // start the first child
+  CHECK(called1);
+  sched.start_next(); // start the second child; this will call set_done
+  CHECK_FALSE(called3);
+  sched.start_next(); // start the third child
+  // TODO: the third child sender should have been canceled
+  // CHECK_FALSE(called3);
+  // CHECK(cancelled);
```

Comment:
I think this doesn't work because `start` on `impulse_scheduler`'s operation state doesn't check the receiver's stop token. It should call `set_done` if stop has been requested.

Reply:
Hehe. I expected magic inside `when_all`, and magic doesn't simply happen :). Fixed it. Thank you!